packages/expo-av/ios/EXAV/AudioSampleCallback/AudioSampleCallbackWrapper.h

```diff
@@ -0,0 +1,46 @@
+// Copyright 2017-present 650 Industries. All rights reserved.
+#pragma once
+
+#import <AVFoundation/AVFoundation.h>
+
+#import <jsi/jsi.h>
+
+// TODO: Replace this with <ReactCommon/TurboModuleUtils.h> after we upgrade to RN 0.66
+#import <EXAV/ReactCallbackWrapper.h>
+
+namespace jsi = facebook::jsi;
+using CallInvoker = facebook::react::CallInvoker;
+
+// TODO: Replace this with just facebook::react namespace after we upgrade to RN 0.66
+using JsiCallbackWrapper = expo::facebook::react::CallbackWrapper;
+using LongLivedObjectCollection = expo::facebook::react::LongLivedObjectCollection;
+
+namespace expo {
+namespace av {
+
+// A class managing lifecycle of audio sample buffer callbacks
+class AudioSampleCallbackWrapper
+{
+  std::weak_ptr<JsiCallbackWrapper> weakWrapper;
+public:
+  AudioSampleCallbackWrapper(jsi::Function &&callback,
+                             jsi::Runtime &runtime,
+                             std::shared_ptr<CallInvoker> jsInvoker);
+  
+  ~AudioSampleCallbackWrapper();
+  
+  void call(AudioBuffer* buffer, double timestamp);
+  
+  // static members
+public:
+  // called when JS VM is destroyed to remove all JSI callback objects
+  static void removeAllCallbacks() {
+    callbackCollection->clear();
+  }
+  
+private:
+  static std::shared_ptr<LongLivedObjectCollection> callbackCollection;
```

Comment:
I know it's a singleton. The problem was more complex here and I deliberately created a separate object here.
- Before RN 0.66 the TurboModules were using this singleton e.g. for promise callbacks.
- In RN 0.66 they replaced the singleton with separate instance: https://github.com/facebook/react-native/commit/32bfd7a857c23dd417f940d0c09843de257f6c61 - that's why I copied some of the RN code here.
- If I were using the singleton, the collection would be shared with non expo-av places, e.g. RN internals and I don't want my code to affect them. 
- The main problem is the global singleton collection is never cleared in main RN code - so the JSC complained about dangling objects (similar issue: https://github.com/software-mansion/react-native-reanimated/issues/1424). So I want only to clear my own collection when `jsc::Runtime` is destructed to avoid potential conflicts with recently quick evolution of RN code.

Reply:
got it. thanks for your explicitly explanation 🔥